stdlib/public/core/TemporaryAllocation.swift

```diff
@@ -0,0 +1,217 @@
+///
+/// - Returns: Whether or not there is sufficient space on the stack to allocate
+///   `byteCount` bytes of memory.
+@_alwaysEmitIntoClient @inline(__always)
+internal func _isStackAllocationSafe(_ byteCount: Int) -> Bool {
+  // Fast path, with a limit known at compile-time. Allocations smaller than
+  // this limit are reasonable to allocate on the stack without worrying about
+  // running out of space, and the compiler would emit such allocations on the
+  // stack anyway when they represent structures or stack-promoted objects.
+  if byteCount <= 1024 {
+    return true
+  }
+
+  // Slower path, with a limit known at runtime and requiring a new standard
+  // library to check.
+  guard #available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *) else {
+    return false
+  }
+  return _swift_stdlib_isStackAllocationSafe(byteCount)
+}
+
+/// Provides scoped access to a raw buffer pointer with the specified byte count
+/// and alignment, allocated on the stack.
+///
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
+///   - body: A closure to invoke and to which the allocated buffer pointer
+///     should be passed. If the buffer pointer could not be allocated, `nil` is
+///     passed to this closure.
+///
+/// - Returns: Whatever is returned by `body`.
+///
+/// - Throws: Whatever is thrown by `body`.
+///
+/// This function encapsulates the various calls to builtins required by
+/// `withUnsafeTemporaryAllocation()`.
+@_alwaysEmitIntoClient @inline(__always)
+internal func _withUnsafeStackAllocation<R>(
+  byteCount: Int,
+  alignment: Int,
+  _ body: (Builtin.RawPointer?) throws -> R
+) rethrows -> R {
+  guard _isStackAllocationSafe(byteCount) else {
+    return try body(nil)
+  }
+
+  // This declaration must come BEFORE Builtin.stackAlloc() or
+  // Builtin.stackDealloc() will end up blowing it away (and the verifier will
+  // notice and complain.)
+  let result: R
+  
+  let stackAddress = Builtin.stackAlloc(
+    byteCount._builtinWordValue,
+    alignment._builtinWordValue
+  )
+  // The multiple calls to Builtin.stackDealloc() are because defer { } produces
+  // a child function at the SIL layer and that conflicts with the verifier's
+  // idea of a stack allocation's lifetime.
+
+  do {
+    result = try body(stackAddress)
+    Builtin.stackDealloc(stackAddress)
+    return result
+
+  } catch {
+    Builtin.stackDealloc(stackAddress)
+    throw error
+  }
+}
+
+// MARK: - Public interface
+
+/// Provides scoped access to a raw buffer pointer with the specified byte count
+/// and alignment.
+///
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
```

Comment:
Sorry for the rehash, but if we don't state a power-of-2 requirement here, then maybe the actual behaviour should be described. The power-of-2 requirement for UMRP is stated [here](https://github.com/apple/swift/blob/fb8ec6faf66ec88581f5b14b29b3f7b3811ea6f9/stdlib/public/core/UnsafeRawPointer.swift#L603), and similarly for UMRBP.